lua/ttt2/libraries/modelbuilder.lua

```diff
@@ -0,0 +1,316 @@
+		local bone = viewModel:LookupBone(identifier)
+
+		if not bone then continue end
+
+		local s = Vector(dataTable.scale.x, dataTable.scale.y, dataTable.scale.z)
+		local p = Vector(dataTable.pos.x, dataTable.pos.y, dataTable.pos.z)
+		local ms = Vector(1, 1, 1)
+
+		local cur = viewModel:GetBoneParent(bone)
+
+		while (cur >= 0) do
+			local pScale = allBones[viewModel:GetBoneName(cur)].scale
+
+			ms = ms * pScale
+			cur = viewModel:GetBoneParent(cur)
+		end
+
+		s = s * ms
+
+		if viewModel:GetManipulateBoneScale(bone) ~= s then
+			viewModel:ManipulateBoneScale(bone, s)
+		end
+
+		if viewModel:GetManipulateBoneAngles(bone) ~= dataTable.angle then
+			viewModel:ManipulateBoneAngles(bone, dataTable.angle)
+		end
+
+		if viewModel:GetManipulateBonePosition(bone) ~= p then
+			viewModel:ManipulateBonePosition(bone, p)
+		end
+	end
+end
+
+---
+-- Builds a render order or returns the cache if already built. The render order is necessary
+-- because models have to be drawn before sprites and quads.
+-- @param table elements the view elements table
+-- @param table cachedRenderOrder The cached render order that is uesed as a fallback
+-- @return table Returns a new or the cached render order
+-- @realm client
+function modelbuilder.BuildRenderOrder(elements, cachedRenderOrder)
+	if cachedRenderOrder then
+		return cachedRenderOrder
+	end
+
+	local newRenderOrder = {}
+
+	for identifier, dataTable in pairs(elements) do
+		if dataTable.type == "Model" then
+			table.insert(newRenderOrder, 1, identifier)
+		elseif dataTable.type == "Sprite" or dataTable.type == "Quad" then
+			table.insert(newRenderOrder, identifier)
+		end
+	end
+
+	return newRenderOrder
+end
+
+---
+-- Gets the orientation of a given bone.
+-- @param Entity wep The weapon whose bone orientation is desired
+-- @param table baseDataTable The base data table, most of the time the view elements
+-- @param table dataTable The data table, most of the time the model data
+-- @param Entity boneEntity The bone entity, can be the view model, the player or the weapon
+-- @param string boneOverride An override to get a secific bone
+-- @return Vector pos The bone position
+-- @return Angle ang The bone angle
+-- @realm client
+function modelbuilder.GetBoneOrientation(wep, baseDataTable, dataTable, boneEntity, boneOverride)
+	local bone, pos, ang
+
+	if dataTable.rel and dataTable.rel ~= "" then
+		local tbl = baseDataTable[dataTable.rel]
+
+		if not tbl then return end
+
+		-- Technically, if there exists an element with the same name as a bone
+		-- you can get in an infinite loop. Let's just hope nobody's that stupid.
+		pos, ang = modelbuilder.GetBoneOrientation(wep, baseDataTable, tbl, boneEntity)
+
+		if not pos then return end
+
+		pos = pos + ang:Forward() * tbl.pos.x + ang:Right() * tbl.pos.y + ang:Up() * tbl.pos.z
+
+		ang:RotateAroundAxis(ang:Up(), tbl.angle.y)
+		ang:RotateAroundAxis(ang:Right(), tbl.angle.p)
+		ang:RotateAroundAxis(ang:Forward(), tbl.angle.r)
```

Comment:
This seems pretty inconsistent to me, why are these rotations etc only applied to the result if it is coming from the "rel" lookup.

This whole method really seems very specifically crafted for another function instead of some public available function. If so, add comments and make this local

Reply:
Honestly, I don't know. This part with the `rel` is not something I used so far. But it is something that was done by original addon and I wanted to support it, so I kept it.

Reply:
Weird stuff, but we surely can refactor this. Either remove these inconsistencies and try to apply those after the function call if possible or we make this a local function instead of some module function then i wouldn't care.

Reply:
made it local